### PR TITLE
WIP: Use jumptables for variants/enums when possible

### PR DESCRIPTION
**Pull request by hurryabit**
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

---
This change is [Reviewable](https://reviewable.io/reviews/digital-asset/daml/5794)

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 1]
```diff
@@ -576,8 +576,9 @@ object Speedy {
     }
   }
 
-  /** The scrutinee of a match has been evaluated, now match the alternatives against it. */
-  final case class KMatch(alts: Array[SCaseAlt]) extends Kont with SomeArrayEquals {
+  /** The scrutinee of a match has been evaluated, now match the alternatives against it.
+    * Consult the documentation of `SECase` for the meaning of `jumpable`. */
+  final case class KMatch(alts: Array[SCaseAlt], jumpable: Boolean) extends Kont with SomeArrayEquals {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:581]: Why not create specialize version of `KMacth`  
why not specializing  `KMach` for each matchable type ? 

```[scala]
final case class KMatchBoolean(ifCase: SExpr, elseCase: SExpr) extends Kont with SomeArrayEquals {
    override def execute(v: SValue, machine: Machine): Unit =
      machine.ctrl = CtrlExpr(if (v.asInstanceOf[SBool].value) ifCase else elseCase)
  }
  
  final case class KMatchEnum(alts: Array[SExpr]) extends Kont with SomeArrayEquals {
    override def execute(v: SValue, machine: Machine): Unit =
      machine.ctrl = CtrlExpr(alts(v.asInstanceOf[SEnum].constructorRank))
  }

  final case class KMatchVariant(alts: Array[SExpr]) extends Kont with SomeArrayEquals {
    override def execute(v: SValue, machine: Machine): Unit = {
      val variant = v.asInstanceOf[SVariant]
      machine.kont.add(KPop(1))
      machine.env.add(variant.value)
      machine.ctrl = CtrlExpr(alts(variant.constructorRank))
    }
  }
  ...

 ```

**Comment by nickchapman-da** [comment]
Do you get any speed up from this?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 1 above)
Comment (hurryabit) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:581]: I'm working on something similar in a more general fashion.

**Comment by hurryabit** [comment]
I've seen some small speedups but they might also be noise. Looks like we need to fry bigger fish before this makes an actual impact.

**Review by nickchapman-da — COMMENTED** [review]
no comment

**Comment by garyverhaegen-da** [comment]
Open for almost a year and no recent movement: closing. If anyone is still actively working on this, please open a new PR.

**Pull request closed without merging**